internal/manifest/manifest.go

```diff
@@ -0,0 +1,250 @@
+package manifest
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"slices"
+
+	"github.com/klauspost/compress/zstd"
+
+	"github.com/canonical/chisel/internal/jsonwall"
+	"github.com/canonical/chisel/internal/setup"
+)
+
+const Filename = "manifest.wall"
+const Schema = "1.0"
+const Mode fs.FileMode = 0644
+
+type Package struct {
+	Kind    string `json:"kind"`
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Digest  string `json:"sha256"`
+	Arch    string `json:"arch"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path"`
+	Mode      string   `json:"mode"`
+	Slices    []string `json:"slices"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice"`
+	Path  string `json:"path"`
+}
+
+type Manifest struct {
+	db *jsonwall.DB
+}
+
+func Read(absPath string) (manifest *Manifest, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot read manifest: %s", err)
+		}
+	}()
+
+	file, err := os.OpenFile(absPath, os.O_RDONLY, Mode)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	r, err := zstd.NewReader(file)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Close()
+	jsonwallDB, err := jsonwall.ReadDB(r)
+	if err != nil {
+		return nil, err
+	}
+
+	manifest = &Manifest{db: jsonwallDB}
+	err = validate(manifest)
```

Comment:
[Question for reviewer]: Is it worth it to validate all the manifests? That means reading all the contents and un-marshalling them. The other option is that this function is made public (`Validate`) and is only called at the user discretion. In our case we will call it from the tests just to make sure that everything is correct.

Reply:
Yeah, your intuition is right on that one: it definitely doesn't make sense to linearly scan everything and cross-check the data on a structure that is meant to perform well on random access. The validation function might come handy, but should not be on the fast path.

Reply:
I have made `Validate` a public function instead and now it is out of the hot path. I have also added some comments to the functions to state that we are NOT validating the manifest.

All was done as part of https://github.com/canonical/chisel/pull/144/commits/401c86e70f3682ad4c01abf9a173f9fac32de20d.